packages/crypto/src/bls12_318/pairing.rs

```diff
@@ -0,0 +1,234 @@
+use crate::{
+    errors::AggregationPairingEquality, CryptoError, BLS12_381_G1_POINT_LEN, BLS12_381_G2_POINT_LEN,
+};
+
+use super::points::{g1_from_variable, g2_from_variable};
+use ark_bls12_381::Bls12_381;
+use ark_ec::{
+    bls12::{G1Prepared, G2Prepared},
+    pairing::Pairing,
+};
+use num_traits::Zero;
+use rayon::iter::{ParallelBridge, ParallelIterator};
+
+pub fn bls12_381_aggregate_pairing_equality(
+    ps: &[u8],
+    qs: &[u8],
+    r: &[u8],
+    s: &[u8],
+) -> Result<bool, CryptoError> {
+    if ps.is_empty() {
+        return Err(AggregationPairingEquality::EmptyG1.into());
+    } else if qs.is_empty() {
+        return Err(AggregationPairingEquality::EmptyG2.into());
+    } else if ps.len() % BLS12_381_G1_POINT_LEN != 0 {
+        return Err(AggregationPairingEquality::NotMultipleG1 {
+            remainder: ps.len() % BLS12_381_G1_POINT_LEN,
+        }
+        .into());
+    } else if qs.len() % BLS12_381_G2_POINT_LEN != 0 {
+        return Err(AggregationPairingEquality::NotMultipleG2 {
+            remainder: qs.len() % BLS12_381_G2_POINT_LEN,
+        }
+        .into());
+    } else if (ps.len() / BLS12_381_G1_POINT_LEN) != (qs.len() / BLS12_381_G2_POINT_LEN) {
+        return Err(AggregationPairingEquality::UnequalPointAmount {
+            left: ps.len() / BLS12_381_G1_POINT_LEN,
+            right: qs.len() / BLS12_381_G2_POINT_LEN,
+        }
+        .into());
+    }
+
+    let pq_pairs: Vec<_> = ps
+        .chunks_exact(BLS12_381_G1_POINT_LEN)
+        .zip(qs.chunks_exact(BLS12_381_G2_POINT_LEN))
+        // From here on parallelism is fine since the miller loop runs over
+        // a sum of the pairings and is therefore a commutative operation
+        .par_bridge()
+        .map(|(p, q)| {
+            let g1 = g1_from_variable(p)?;
+            let g2 = g2_from_variable(q)?;
+
+            Ok((G1Prepared::from(g1.0), G2Prepared::from(g2.0)))
+        })
+        .collect::<Result<_, CryptoError>>()?;
+
+    let r = g1_from_variable(r)?;
+    let s = g2_from_variable(s)?;
+
+    let r_neg = G1Prepared::from(-r.0);
+    let s_prepared = G2Prepared::from(s.0);
+
+    let (ps, qs): (Vec<_>, Vec<_>) = pq_pairs.into_iter().chain([(r_neg, s_prepared)]).unzip();
+
+    Ok(Bls12_381::multi_pairing(ps, qs).is_zero())
+}
+
+/// Check whether the following condition holds true:
+///
+/// $$
+/// e(p, q) = e(r, s)
+/// $$
+pub fn bls12_381_pairing_equality(
```

Comment:
Is there a good reason to keep this if we have bls12_381_aggregate_pairing_equality?

Reply:
Not really, no. Gonna remove that.

Reply:
One argument one could have is not having to transform the points in the caller code (p to -1). But then again whoever touches this tech probably has harder issues to solve.

Reply:
Actually, since we invert the `r` component of the aggregation equality, end users still don't have to negate anything themselves.  
Because it doesn't matter whether `p` or `r` is negated, the result will be the same.

Reply:
Uhh, I realized we are doing something smart here: https://hackmd.io/@webmaster128/r18mwFpbA

Reply:
Yeah, we do this to save on the amount of final exponentiations (reducing it to a single final exponentiation), which should yield considerable time savings.

The reason why I chose to do that inversion in the host code is because I really didn't want to make the consumer of the API have to think about inversions, etc.

They just pass in the points as-is (as they would expect from RFC descriptions) and get a Boolean back. Simple.

Reply:
I'm asking some cryptographer friends now but I think this is the way. It's nicer than the Ethereum API in terms of API design.

Reply:
Okay, this is good. Everything confirmed.

Now one advantage the Ethereum API has is you can test `e(a, b) = 1` because all the terms are on the left side. We can achieve the same by allowing `n = 0` on the left side. This would then automatically translate to `1 = e(r, s)`. Not sure how common this case is, but I guess it's good to have options.

Gas-wise we whis would lead to cases with 1 or more pairings plus the final exponantiation. So I guess this needs to be adjusted a bit.

Reply:
By what factor would you adjust the gas values?

Reply:
I think the formula should be: `k * c_1 + c_2` where k is the number if pairings (`k = n+1)`. The constants c_1 and c_2 can then be filled with values obtained by benchmarking.

Reply:
We pretty much have that cost calculation in place already, except we subtract one since technically the pairing equality benchmark benched $e(p, q) = e(r, s)$

Therefore we already have the equal number of pairings if we subtract one since the base already contains two pairings.

So doing $((n - 1) * C_{pp}) + C_{base}$ should be fine, no?

Reply:
This is fine as long as `n >= 1`. But being able to set `n = 0`/`k=1` would be good. In this case the `saturating_sub` is losing information.

But you can easily take the existing constant `bls12_381_pairing_equality_cost` and subtract two pairings from it.